src/she/sdl2/sdl2_surface.cpp

```diff
@@ -0,0 +1,431 @@
+    return from_sdl(m_bmp->format, data);
+  }
+
+  void SDL2Surface::putPixel(gfx::Color color, int x, int y)
+  {
+    auto clip = m_bmp->clip_rect;
+    if (x < clip.x || x >= clip.x + clip.w || y < clip.y || y >= clip.y + clip.h)
+      return;
+    int sdlColor = to_sdl(m_bmp->format, color);
+    if (m_bmp->format->BytesPerPixel == 4)
+      *reinterpret_cast<uint32_t*>(getData(x, y)) = sdlColor;
+    else if (m_bmp->format->BytesPerPixel == 2)
+      *reinterpret_cast<uint16_t*>(getData(x, y)) = sdlColor;
+    else if (m_bmp->format->BytesPerPixel == 1)
+      *reinterpret_cast<uint8_t*>(getData(x, y)) = sdlColor;
+  }
+
+  void SDL2Surface::drawHLine(gfx::Color color, int x, int y, int w)
+  {
+    SDL_Rect clip;
+    SDL_GetClipRect(m_bmp, &clip);
+
+    if (x < clip.x) {
+      w += x - clip.x;
+      x = clip.x;
+    }
+
+    if (x + w >= clip.x + clip.w) {
+      w = (clip.x + clip.w) - x - 1;
+    }
+
+    if (w <= 0 || y < clip.y || y >= (clip.y + clip.h)) {
+      return;
+    }
+
+    int sdlColor = to_sdl(m_bmp->format, color);
+    auto data = getData(x, y);
+    if (m_bmp->format->BytesPerPixel == 4) {
+      for(; w--; data += 4)
+        *reinterpret_cast<uint32_t*>(data) = sdlColor;
+    } else if (m_bmp->format->BytesPerPixel == 2) {
+      for(; w--; data += 2)
+        *reinterpret_cast<uint16_t*>(data) = sdlColor;
+    } else if (m_bmp->format->BytesPerPixel == 1) {
+      for(; w--; data += 1)
+        *reinterpret_cast<uint8_t*>(data) = sdlColor;
+    }
+  }
+
+  void SDL2Surface::drawVLine(gfx::Color color, int x, int y, int h)
+  {
+    SDL_Rect clip;
+    SDL_GetClipRect(m_bmp, &clip);
+
+    if (y < clip.y) {
+      h += y - clip.y;
+      y = clip.y;
+    }
+    if (y + h >= clip.y + clip.h) {
+      h = (clip.y + clip.h) - y - 1;
+    }
+    if (h <= 0 || x < clip.x || x >= clip.x + clip.w) {
+      return;
+    }
+    int sdlColor = to_sdl(m_bmp->format, color);
+    auto data = getData(x, y);
+    auto stride = m_bmp->pitch;
+    if (m_bmp->format->BytesPerPixel == 4) {
+      for(; h--; data += stride)
+        *reinterpret_cast<uint32_t*>(data) = sdlColor;
+    } else if (m_bmp->format->BytesPerPixel == 2) {
+      for(; h--; data += stride)
+        *reinterpret_cast<uint16_t*>(data) = sdlColor;
+    } else if (m_bmp->format->BytesPerPixel == 1) {
+      for(; h--; data += stride)
+        *reinterpret_cast<uint8_t*>(data) = sdlColor;
+    }
+  }
+
+  void SDL2Surface::drawLine(gfx::Color color, const gfx::Point& a, const gfx::Point& b)
+  {
+    if (a.x == b.x) {
+      int y = a.y;
+      int h = b.y - a.y;
+      if (h < 0) {
+        y = b.y;
+        h = -h;
+      }
+      drawVLine(color, a.x, y, h);
+      return;
+    }
+
+    if (a.y == b.y) {
+      int x = a.x;
+      int w = b.x - a.x;
+      if (w < 0) {
+        x = b.x;
+        w = -w;
+      }
+      drawHLine(color, x, a.y, w);
+      return;
+    }
+
+    printf("Unsupported line: %d,%d -> %d,%d\n", a.x, a.y, b.x, b.y);
```

Comment:
This needs to be implemented. Perhaps [Bresenham's line algorithm](https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm) can be used.

Reply:
I'm guessing there must already be an implementation of it somewhere within the code. It would be very odd if a drawing app didn't have a line algorithm. Maybe just need to find and call it?

Reply:
I think that Allegro's drawing functions were used previously. Allegro 4 provides a `line` function to draw lines, but as far as I know SDL2 doesn't have anything equivalent.

Reply:
https://wiki.libsdl.org/SDL_RenderDrawLine ?

Reply:
That only works for hardware-accelerated textures, not Surfaces.